src/source/source_state.js

```diff
@@ -16,50 +16,144 @@ export type LayerFeatureStates = {[layer: string]: FeatureStates};
+
+
+        for (const key in newState) {
+            const deletionInQueue = this.deletedStates && this.deletedStates[sourceLayer] && this.deletedStates[sourceLayer][feature] && this.deletedStates[sourceLayer][feature][key] === null;
+            if (deletionInQueue) {
+                delete this.deletedStates[sourceLayer][feature][key];
+            }
+        }
+    }
+
+    removeFeatureState(sourceLayer: string, featureId?: number, key?: string) {
+
+        const feature = String(featureId);
+
+        this.deletedStates = this.deletedStates || {};
+        this.deletedStates[sourceLayer] = this.deletedStates[sourceLayer] || {};
+
+
+        if (key && featureId) {
+            this.deletedStates[sourceLayer][feature] = this.deletedStates[sourceLayer][feature] || {};
+            this.deletedStates[sourceLayer][feature][key] = null;
+        } else if (featureId) {
+            this.deletedStates[sourceLayer][feature] = {};
+
+            const featureStateExists = this.state[sourceLayer] && this.state[sourceLayer][feature];
+            const updateInQueue = this.stateChanges[sourceLayer] && this.stateChanges[sourceLayer][feature];
+
+            if (featureStateExists) {
+                this.deletedStates[sourceLayer][feature] = null;
+            }
+
+            if (updateInQueue) {
+
+                for (key in this.stateChanges[sourceLayer][feature]) {
+                    this.deletedStates[sourceLayer][feature][key] = null;
+                }
+            }
+
+        } else {
+            this.deletedStates[sourceLayer] = {};
+
+            const featureStateExists = this.state[sourceLayer];
+            const updateInQueue = this.stateChanges[sourceLayer];
+
+            if (featureStateExists) {
+                for (const feature in featureStateExists) {
+                    this.deletedStates[sourceLayer][feature] = {};
+                    for (key in this.state[sourceLayer][feature]) {
+                        this.deletedStates[sourceLayer][feature][key] = null;
+                    }
+                }
+            }
+
+            if (updateInQueue) {
+                for (const feature in updateInQueue) {
+                    this.deletedStates[sourceLayer][feature] = {};
+                    for (key in this.stateChanges[sourceLayer][feature]) {
+                        this.deletedStates[sourceLayer][feature][key] = null;
+                    }
+                }
+            }
+        }
     }
 
     getState(sourceLayer: string, featureId: number) {
         const feature = String(featureId);
         const base = this.state[sourceLayer] || {};
         const changes = this.stateChanges[sourceLayer] || {};
-        return extend({}, base[feature], changes[feature]);
+        const deletions = this.deletedStates && this.deletedStates[sourceLayer] ? this.deletedStates[sourceLayer] : {};
+        return extend({}, base[feature], changes[feature], deletions[feature]);
```

Comment:
@asheemmamoowala running into a surprising snag here when trying to delete state keys instead of setting them to null. The getState() method essentially has to run a coalesce of its own to ensure that it's returning a reconciled version of state, updates, and deletions. Deletions are represented by `null`s, and applied to the state via `extend()`. 

Unfortunately, the `const` syntax (immutable?) is preventing me from deleting the nulls from the `extend`ed object. I can declare another root-level variable alongside `state`, `stateChanges`, and `deletedStates`, or write a custom version of `extend()` that can actually delete stuff, but both of these feel like overkill